File: MetalHello/MetalHello/ViewController.m
//
//  ViewController.m
//  MetalHello
//
//  Created by 余河川 on 2018/11/12.
//  Copyright © 2018 余河川. All rights reserved.
//

#import "ViewController.h"
#import <MetalKit/MetalKit.h>
#import <QuartzCore/QuartzCore.h>

@interface ViewController ()

/**
 <#Description#>
 */
@property (nonatomic, strong) MTKView *mtkView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    id<MTLDevice> device = MTLCreateSystemDefaultDevice();
    if (device == nil) {
        NSLog(@"不支持的设备");
        return ;
    }
    
    self.mtkView = [[MTKView alloc] initWithFrame:CGRectZero device:device];
    [self.view addSubview:_mtkView];
    _mtkView.frame = self.view.bounds;
    
    id<MTLCommandQueue> queue = device.newCommandQueue;
    
    id<MTLCommandBuffer> buffer = [queue commandBuffer];
    
    
    static const float vertexArrayData[] = {
        
        0.577, -0.25, 0.0, 1.0,
        -0.577, -0.25, 0.0, 1.0,
        0.0, 0.5, 0.0, 1.0
        
    };
    
    id<MTLBuffer> vertexBuffer = [device newBufferWithBytes:vertexArrayData length:sizeof(vertexArrayData) options:0];
    
    UIImage *image = [UIImage imageNamed:@"orgin"];
    MTKTextureLoader *loader = [[MTKTextureLoader alloc] initWithDevice:device];
    NSError *err = nil;
    id<MTLTexture> texture = [loader newTextureWithCGImage:image.CGImage options:nil error:&err];
    
    MTLRenderPipelineDescriptor *des = [MTLRenderPipelineDescriptor new];
    id<MTLLibrary> library = [device newDefaultLibrary];
    des.vertexFunction =  [library newFunctionWithName:@"myVertexShader"];
    des.fragmentFunction = [library newFunctionWithName:@"myFragmentShader"];
    des.colorAttachments[0].pixelFormat = MTLPixelFormatBGRA8Unorm;
    
    NSError *error;
    id <MTLRenderPipelineState> pipelineState = [device newRenderPipelineStateWithDescriptor:des error:&error];
    
    CAMetalLayer *metalLayer = (CAMetalLayer*)[_mtkView layer];
    id<CAMetalDrawable> drawable = [metalLayer nextDrawable];
    
    MTLRenderPassDescriptor *renderDes = [MTLRenderPassDescriptor new];
    renderDes.colorAttachments[0].texture = drawable.texture;
    renderDes.colorAttachments[0].loadAction = MTLLoadActionClear;
    renderDes.colorAttachments[0].storeAction = MTLStoreActionStore;
    renderDes.colorAttachments[0].clearColor = MTLClearColorMake(0.5, 0.65, 0.8, 1); //background color
    
    
    //command encoder
    id<MTLRenderCommandEncoder> encoder = [buffer renderCommandEncoderWithDescriptor:renderDes];
    [encoder setCullMode:MTLCullModeNone];
    [encoder setFrontFacingWinding:MTLWindingCounterClockwise];
    [encoder setRenderPipelineState:pipelineState];
    [encoder setVertexBuffer:vertexBuffer offset:0 atIndex:0];
    [encoder setFragmentTexture:texture atIndex:0];
    
    //set render vertex
    [encoder drawPrimitives:MTLPrimitiveTypeTriangle
                vertexStart:0
                vertexCount:3];
    
    [encoder endEncoding];
    
    //commit
    [buffer presentDrawable:drawable];
    [buffer commit];
    
}




@end

File: Metal181116_2/Metal181116_2/ViewController.swift
//
//  ViewController.swift
//  Metal181116_2
//
//  Created by 余河川 on 2018/11/16.
//  Copyright © 2018 余河川. All rights reserved.
//

import UIKit
import MetalKit

class ViewController: UIViewController, MTKViewDelegate {
    

    
    public var device: MTLDevice!
    var queue: MTLCommandQueue!
    var cps: MTLComputePipelineState!
    let mtkView = MTKView.init(frame: CGRect.zero)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.view.addSubview(mtkView)
        self.mtkView.delegate = self
        device = MTLCreateSystemDefaultDevice()!
        self.mtkView.delegate = self
        self.mtkView.device = self.device
        self.mtkView.framebufferOnly = false
        registerShaders()
        
        
    }
    
    override func viewWillLayoutSubviews() {
        self.mtkView.frame = self.view.bounds
    }
    
    func registerShaders() {
        device = MTLCreateSystemDefaultDevice()!
        queue = device.makeCommandQueue()
        let library = device.makeDefaultLibrary()
        let kernel = library?.makeFunction(name: "compute")!
        cps = try! device.makeComputePipelineState(function: kernel!)
    }
    
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        mtkView.isPaused = false
    }
    
    func draw(in view: MTKView) {
        
        if let drawable = view.currentDrawable,
            let commandBuffer = queue.makeCommandBuffer(),
            let commandEncoder = commandBuffer.makeComputeCommandEncoder() {
            commandEncoder.setComputePipelineState(cps)
            commandEncoder.setTexture(drawable.texture, index: 0)
            let threadGroupCount = MTLSizeMake(8, 8, 1)
            let threadGroups = MTLSizeMake(drawable.texture.width / threadGroupCount.width, drawable.texture.height / threadGroupCount.height, 1)
            commandEncoder.dispatchThreadgroups(threadGroups, threadsPerThreadgroup: threadGroupCount)
            commandEncoder.endEncoding()
            commandBuffer.present(drawable)
            commandBuffer.commit()
            mtkView.isPaused = true
        }
        
    }


}

